baybe/constraints/discrete.py

```diff
@@ -41,6 +43,17 @@ def get_invalid(self, data: pd.DataFrame) -> pd.Index:  # noqa: D102
         res = reduce(_valid_logic_combiners[self.combiner], satisfied)
         return data.index[res]
 
+    def get_polars_expression(self, df: pl.LazyFrame) -> pl.Expr:
+        """Translate the conditions to polars expression for filtering.
+
+        Args:
+            df: A polars lazy frame of all parameters.
+
+        Returns:
+            The Polars Expr object to pass as an argument to filter().
+        """
+        return pl.lit(True)
```

Comment:
I don't understand this part, tbh. You added `pl.lit(True)` as a mypy fix, but returning this expression is semantically not identical to executing `pass`. Haven't stepped through the code yet, but shouldn't this part actually implement the logic of the constraints in polars!? Same for the other constraints, of course... Could it be that what you actually meant is a `NotImplementedError`?

Reply:
if we decide that not all constraints need the polars functionality this and other `pl.lit` implementations would vanish also vanishing this question

Reply:
>  Could it be that what you actually meant is a NotImplementedError?

Yes, I used `pl.lit(True)` for the same reason.
I'm gonna with @Scienfitz suggestion to remove unnecessary `get_polars_expression`s

Reply:
what do you mean didnt work? This suggestion was not related to typing issues

Reply:
When I remove it from `DiscreteConstraint`,   the following lines of searchspace.discrete.py become problematic:
```
for c in constraints:
            pl_expr = c.get_polars_expression(res)
            if pl_expr is not None:
                res = res.filter(pl_expr)
```
There's a mypy error
```
baybe/searchspace/discrete.py:749: error: "DiscreteConstraint" has no attribute "get_polars_expression"  [attr-defined]
```
Is it ok to change the input type to `constraints: Iterable[Union[DiscreteSumConstraint, DiscreteProductConstraint, DiscreteExclusionConstraint]]`?

Reply:
Ok, just for clarification. The error is because INSIDE the method, it is not clear that the list elements of `constraints` have the `get_polars_expression` method implement, hence mypy complains. In order for that to happen, some form of [type narrowing](https://mypy.readthedocs.io/en/stable/type_narrowing.html) must happen, which in principle could work through some form of `assert hasattr(x, "get_polars_expression")` check. You indeed have such a check, namely in the list comprehension for `polars_constraints` in the calling function where you subselect the objects that have this method implemented. However, when being passed to `parameter_cartesian_prod_polars`, this is overshadowed by the signature of that method. 

So in order to resolve it, there are only two ways:
* We ignore it.
* We apply some form of type narrowing that mypy can see.

The possible paths I can see for the latter:
- Your suggestion would solve the problem but has the downside of maintaining that new union type.
- We define some type that indicates that the object has the method implemented, e.g. a `DiscreteConstraintWithPolarsExpr` as a subclass of `DiscreteConstraint` (but I'm not a big fan of having super cluttered class hierarchies)
- We somehow make it cleverly work using a protocol that can be employed in the signature of the function
- ...

However, all my ideas seem more complicated that a simple ignore here. @Scienfitz, any other suggestions?

Reply:
youre still not doing any type narrowing a la `hasattr` or similar right? We have mypy ignores for such situations in many places

 imo thats the simplest way and fine, ideally the line that does the type narrow should be as close as possible to the line with the ignore, that would mean directly inside `parameter_cartesian_prod_polars ` before you iterate over the constraints, no union needed